Source/WebCore/page/scrolling/ScrollSnapOffsetsInfo.cpp

```diff
@@ -347,7 +347,7 @@ void updateSnapOffsetsForScrollableArea(ScrollableArea& scrollableArea, const Re
     auto scrollSnapPort = computeScrollSnapPortOrAreaRect(viewportRectInBorderBoxCoordinates, scrollingElementStyle.scrollPadding(), InsetOrOutset::Inset);
     LOG_WITH_STREAM(ScrollSnap, stream << "Computing scroll snap offsets for " << scrollableArea << " in snap port " << scrollSnapPort);
     for (auto& child : boxesWithScrollSnapPositions) {
-        if (child.enclosingScrollableContainerForSnapping() != &scrollingElementBox)
+        if (child.enclosingScrollableContainerForSnapping() != &scrollingElementBox || !child.element())
```

Comment:
Aside from resolving the crash, can you explain why it's correct that we skip this work when `child.element()` is null? It could be null when the box is for generated content like a `::before`. Is it right that we skip updating snap offsets for generated content? (Do we support scroll snapping to generated content? The example at https://drafts.csswg.org/css-scroll-snap-1/#re-snap suggests that we should, but I don't know if we do.)

Reply:
I might be wrong, and we do return a non-null element here for generated content. So maybe it is for anonymous content? In which case it is fine? Either way, please do explain in the commit message why it is OK to skip in this case.

Reply:
Hmm yeah I haven't enumerated all of the cases where we have a renderer but not an element, so I'm not sure if it's ok in all cases. I'll try to make some tests around this.

Reply:
Confirmed that we do generate an element in this case, so it would only be null for anonymous renderers.

Reply:
Doesn't `renderer->element()` return the same thing for a pseudo's render and the element's normal renderer? If so, how can you distinguish between snapping to the element, and snapping to the pseudo?

Reply:
No renderer->element() returns a generated node with type pseudo, with its own unique identifier from the element's renderer.

Reply:
It would be good to add a test case that snaps to both elements and their `::before` and `::after` so we prove that those are individual snap targets.

Reply:
Unfortunately I don't think there's a great way to force snapping to both `::before` and `::after` elements, since the behavior for choosing between them is essentially random (in `chooseBoxToResnapTo` we just choose the first element in snappedBoxes).